src/node/stage/approved.rs

```diff
@@ -1244,22 +1291,153 @@ impl Approved {
         &mut self,
         core: &mut Core,
         elders_info: EldersInfo,
-        section_key: bls::PublicKey,
-        proof: AccumulatingProof,
+        proof: Proof,
     ) -> Result<()> {
+        let elders_info = Proven::new(elders_info, proof);
+
+        if elders_info.value.prefix == *self.shared_state.our_prefix()
+            || elders_info
+                .value
+                .prefix
+                .is_extension_of(self.shared_state.our_prefix())
+        {
+            // Our section
+            if let Some(details) = self.section_update_barrier.handle_info(
+                core.name(),
+                self.shared_state.our_prefix(),
+                elders_info,
+            ) {
+                self.update_our_section(core, details)?
+            }
+        } else {
+            // Other section
+            self.update_neighbour_info(core, elders_info)
+        }
+
+        Ok(())
+    }
+
+    fn handle_send_neighbour_info_event(
+        &mut self,
+        core: &mut Core,
+        dst: XorName,
+        nonce: MessageHash,
+    ) -> Result<()> {
+        if !self.is_our_elder(core.id()) {
+            return Ok(());
+        }
+
+        self.send_routing_message(
+            core,
+            SrcLocation::Section(*self.shared_state.our_prefix()),
+            DstLocation::Section(dst),
+            Variant::NeighbourInfo {
+                elders_info: self.shared_state.our_info().clone(),
+                nonce,
+            },
+            None,
+        )
+    }
+
+    fn handle_our_key_event(
+        &mut self,
+        core: &mut Core,
+        prefix: Prefix<XorName>,
+        key: bls::PublicKey,
+        proof: Proof,
+    ) -> Result<()> {
+        if !prefix.matches(core.name()) {
+            return Ok(());
+        }
+
+        let key = Proven::new(key, proof);
+
+        if let Some(details) = self
+            .section_update_barrier
+            .handle_our_key(self.shared_state.our_prefix(), key)
+        {
+            self.update_our_section(core, details)
```

Comment:
looks like this `self.update_our_section` is always inside the barrier check,
then why not have the barrier check inside the `update_our_section` function to reduce duplication?

Reply:
Because there are actually three different barrier checks: `handle_our_key`, `handle_their_key` and `handle_info`. Each is for different event.